### ISSUE TITLE
RoiAlign CPU is not aligned to pixel centers (per the Mask RCNN paper and Facebook's Detectron2 implementation)

### ISSUE THREAD

**Issue by fdwr** [microsoft/onnxruntime#6921]
**Describe the bug**
The [RoiAlign operator](https://github.com/onnx/onnx/blob/master/docs/Operators.md#RoiAlign), per the [Mask RCNN paper](https://arxiv.org/abs/1703.06870) and Facebook Research's [Detectron 2 implementation](https://github.com/facebookresearch/detectron2/blob/master/detectron2/layers/roi_align.py) aligns sampling points over the center of the pixels, but ORT's CPU implementation is half a pixel off for all regions. After comparing ORT to various references (table below), it's evident that the current ORT code duplicated PyTorch's earlier bug in [roi_align](https://github.com/facebookresearch/detectron2/blob/master/detectron2/layers/roi_align.py) which applied an offset the output subsample by 0.5 but forgot to adjust the input sample to compensate. See their comment in the code: "_the original roi_align (aligned=False) does not subtract the 0.5 when computing neighboring pixel indices and therefore it uses pixels with a slightly incorrect alignment (relative to our pixel model) when performing bilinear interpolation_".

This isn't as evident for larger input image regions, where that misalignment becomes less important relative to the overall region size, but it makes quite a difference for smaller regions. Even identity cases (where the region of interest exactly matches the output tensor) are misaligned. e.g. Taking a 2x1 slice of a 4x1 input _should_ yield exactly that input slice, but ORT's result is 0.5 off each element.

    {
      "op_type": "RoiAlign",
      "version": 10,
      "output_height": 1,
      "output_width": 2,
      "sampling_ratio": 0,
      "spatial_scale": 1,
      "mode": "avg",
      "X": [[[[0,1,2,3]]]],
      "rois": [[1,0, 3,1]], // single region, 2 elements wide, 1 tall, sampling the [1,2]
      "batch_indices": [0],
      "Y": [[[[1,2]]]], // <-------- expected output, actual is [1.5, 2.5]
      "T1": "float32",
      "T2": "int64"
    },

**Urgency**
No deadline.

**System information**
- OS Platform and Distribution: NA, but Windows 10 recent selfhost
- ONNX Runtime installed from (source or binary): source
- ONNX Runtime version: 1.7
- Python version: NA
- Visual Studio version (if applicable): VS2019
- GCC/Compiler version (if compiling from source): NA
- CUDA/cuDNN version: NA
- GPU model and memory: NA

**To Reproduce**
- Run the attached model, reduced repro: 
[RoiAlignDetectronTestCase.zip](https://github.com/microsoft/onnxruntime/files/6094576/RoiAlignDetectronTestCase.zip)

**Expected behavior**
Input sample points are aligned rather than 0.5 off each element.

**Screenshots**
e.g.
![image](https://user-images.githubusercontent.com/1809166/110191631-5d82d880-7dde-11eb-938b-da13a22eda97.png)

**Additional context**

This affects the faster_rcnn and mask_rcnn models in WinML, which appear to have been trained with the incorrect alignment in the first place, because CPU is matching the baseline output, whereas DML follows half pixel alignment (matching Detectron 2) and gets different results.

[faster_rcnn opset 10 model mismatches expected output in DirectML](https://microsoft.visualstudio.com/OS/_workitems/edit/31005511/)  
[mask_rcnn opset 10 onnx zoo model fails to evaluate on DirectML](https://microsoft.visualstudio.com/OS/_workitems/edit/31005388/)  

For an example case (adapted from the Detectron test case), and comparison to other framework results:

```
Input Tensor =
          0.0 1.0 2.0 3.0 4.0 5.0
           |.5 |.5 |.5 |.5 |.5 |
    0.0___ |_|_|_|_|_|_|_|_|_|_|
    1.0___[| 0,| 1,| 2,| 3,| 4 |]
    2.0___[|10,<11,|12,>13,|14 |]
    3.0___[|20,<21,|22,>23,|24 |]
    4.0___[|30,|31,|32,|33,|34 |]
    5.0___[|40,|41,|42,|43,|44 |]

Active region of interest = [[1.0, 1.0, 3.0, 3.0]] // a 2x2 window over the input elements [[11,12],[21,22]]
Output tensor size = [4,4]
```

Reference | Output 4x4, up from 2x2   region
-- | --
✔ MaskedRCNN paper / FB Research Detectron 2 | [ 8.25,  8.75,  9.25,  9.75],   [13.25, 13.75, 14.25, 14.75],   [18.25, 18.75, 19.25, 19.75],   [23.25, 23.75, 24.25, 24.75]
✔ DML   ROI_ALIGN 0 | [ 8.25,  8.75,  9.25,  9.75],   [13.25, 13.75, 14.25, 14.75],   [18.25, 18.75, 19.25, 19.75],   [23.25, 23.75, 24.25, 24.75]
✔ `torchvision.ops.roi_align(aligned=True…)` | [ 8.25,  8.75,  9.25,  9.75],   [13.25, 13.75, 14.25, 14.75],   [18.25, 18.75, 19.25, 19.75],   [23.25, 23.75, 24.25, 24.75]
`torchvision.ops.roi_align(aligned=False…)` *deprecated, legacy flag still exists | [13.75, 14.25, 14.75, 15.25],   [18.75, 19.25, 19.75, 20.25],   [23.75, 24.25, 24.75, 25.25],   [28.75, 29.25, 29.75, 30.25]
ONNX Runtime   1.7 RoiAlign CPU | [13.75, 14.25, 14.75, 15.25],   [18.75, 19.25, 19.75, 20.25],   [23.75, 24.25, 24.75, 25.25],   [28.75, 29.25, 29.75, 30.25]
`tf.image.crop_and_resize(…)` *Note boxes are normalized 0 to 1 (so /4 each ROI element) | [11.00, 11.66, 12.33, 13.00],   [17.66, 18.33, 19.00, 19.66],   [24.33, 25.00, 25.66, 26.33],   [31.00, 31.66, 32.33, 33.00]
`tf.image.resize_bilinear(align_corners=True…)` + `tf.slice` | [11.00, 11.66, 12.33, 13.00],   [17.66, 18.33, 19.00, 19.66],   [24.33, 25.00, 25.66, 26.33],   [31.00, 31.66, 32.33, 33.00]
`tf.image.resize_bilinear(align_corners=False…)` + `tf.slice` | [11.00, 11.50, 12.00, 12.50],   [16.00, 16.50, 17.00, 17.50],   [21.00, 21.50, 22.00, 22.50],   [26.00, 26.50, 27.00, 27.50]
✔ `tf.image.resize_bilinear(half_pixel_centers=True…)` + `tf.slice` | [ 8.25,  8.75,  9.25,  9.75],   [13.25, 13.75, 14.25, 14.75],   [18.25, 18.75, 19.25, 19.75],   [23.25, 23.75, 24.25, 24.75]

----------

PyTorch sample code:
```python
# pip install torch==1.7.1+cpu torchvision==0.8.2+cpu torchaudio===0.7.2 -f https://download.pytorch.org/whl/torch_stable.html
import torch
import torchvision
print("PyTorch version:", torch.__version__)

input = [[[[ 0, 1, 2, 3, 4],
            [10,11,12,13,14],
            [20,21,22,23,24],
            [30,31,32,33,34],
            [40,41,42,43,44]]]]
boxes = [[0, 1,1,3,3]]
output_size = [4,4]
aligned=True
#aligned=False
sampling_ratio=1
spatial_scale=1

# https://pytorch.org/vision/0.8/_modules/torchvision/ops/roi_align.html
output = torchvision.ops.roi_align(
    torch.tensor(input, dtype=torch.float),
    torch.tensor(boxes, dtype=torch.float),
    output_size,
    spatial_scale=spatial_scale,
    sampling_ratio=sampling_ratio,
    aligned=aligned
)

torch.set_printoptions(sci_mode=False)
print(input)
print(boxes)
print(output)
```

----------

TensorFlow sample code:
```python
import os
import tensorflow.compat.v1 as tf

input = [[
            [[ 0.], [ 1.], [ 2.], [ 3.], [ 4.]],
            [[10.], [11.], [12.], [13.], [14.]],
            [[20.], [21.], [22.], [23.], [24.]],
            [[30.], [31.], [32.], [33.], [34.]],
            [[40.], [41.], [42.], [43.], [44.]]
        ]]
boxes = [[1/4,1/4,3/4,3/4],[3/4,3/4,4/4,4/4]]
box_indices = [0,0]
crop_size = [4, 4]

print("TensorFlow version:", tf.__version__) # 1.15.0 (cpu/cuda)

output_size = [5*2, 5*2]
resize_output = tf.image.resize_bilinear(tf.constant(input), output_size, align_corners=False, half_pixel_centers=True)
resize_bilinear_slice_output = tf.slice(resize_output, [0,2,2,0], [1,4,4,1])

# crop_and_resize doesn't scale the image boundaries to pixel centers.
method = 'bilinear'
extrapolation_value = 0
crop_and_resize_output = tf.image.crop_and_resize(
    image=tf.constant(input, dtype=tf.float32), # NHWC
    boxes=tf.constant(boxes, dtype=tf.float32),
    box_ind=tf.constant(box_indices, dtype=tf.int32),
    crop_size=tf.constant(crop_size, dtype=tf.int32),
    method=method,
    extrapolation_value=extrapolation_value
)

with tf.Session(config=config) as session:
    with np.printoptions(precision=3, suppress=True):
        print("input:\n", input)
        print("crop_and_resize:\n", session.run(crop_and_resize_output))
        print("resize_bilinear_and_slice:\n", session.run(resize_bilinear_slice_output))
```

**Comment by hariharans29** [comment]
I will investigate this

**Comment by fdwr** [comment]
Thanks Hari. Note that the coordinates of RoiAlign are on an infinite floating point grid, unlike RoiPool (with integers and that weirdo -1 size adjustment). So a ROI (x1,y1,x2,y2) of [1.6, 1.3, 3.3, 3.75] means a region size of width=1.7 (3.3 - 1.6) and height=1.45 (2.75 - 1.3).

![image](https://user-images.githubusercontent.com/1809166/110446130-aef7c580-8073-11eb-8bc8-fd60eb52eac8.png)

**Comment by hariharans29** [comment]
Hi Dwayne,

I investigated this and here are my findings:

1)  In your table above, this doesn't seem right:

  | torchvision.ops.roi_align(aligned=False…)*deprecated, legacy flag still exists | [13.75, 14.25, 14.75, 15.25],[18.75, 19.25, 19.75, 20.25],[23.75, 24.25, 24.75, 25.25],[28.75, 29.25, 29.75, 30.25]
-- | -- | --
??? 🙃 | ONNX Runtime 1.7 CPU EP RoiAlign | [6.1875, 6.75, 6.75, 7.3125],[11.8125, 12.375, 12.375, 12.9375],[11.8125, 12.375, 12.375, 12.9375],[17.4375, 18, 18, 18.5625]

The reason is these results are produced by ORT when the pooling mode is 'max' for the operator. From the looks of it, Torch only seems to support 'avg' pooling. As soon as you change the mode to 'avg', you will see that the results from ORT are matching TorchVision (aligned = False).

So, my conclusion is ORT's CPU implementation (in avg mode) == Legacy Torch ROIAlign (aligned = False)

2) To ensure backwards compatibility, TorchVision seems to have introduced the `aligned` flag and gives the user the option of "fixing the misalignment"(the legacy aligned = False is still the default mode though).  The Detectron project calls this method with aligned = True and by definition, this will lead to diffs wrt to ORT. So, to summarize, ORT's CPU backend implements the legacy mode and DirectML backend (like Detectron) is implementing the "misalignment fixed"  logic. Does this make sense?

Unfortunately, I don't know how to fix this without breaking backwards compatibility. The way I see it, we may have to introduce a new attribute in ONNX for this op (just like TorchVision introduced the `aligned` flag) that would allow the user to pick which implementation they would like. What do you think ?

**Comment by fdwr** [comment]
@hariharans29 : Oops, sorry for the "max" vs "avg" mixup when I recorded that result 😅. I fixed the table above and wording showing it matches deprecated PyTorch behavior.

Yeah, back compat is a concern. Really ONNX should add an attribute to RoiAlign like it did with Resize and default to half_pixels (ONNX convertors from old opsets can set it to the legacy behavior). Thanks for investigating.
